File: go/socket.go
package altengine

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"math/rand/v2"
	"net/url"
	"strings"
	"sync"
	"time"

	"github.com/coder/websocket"
)

// SocketState is the ChannelSocket lifecycle state.
type SocketState string

// Socket states.
const (
	SocketIdle         SocketState = "idle"
	SocketConnecting   SocketState = "connecting"
	SocketOpen         SocketState = "open"
	SocketReconnecting SocketState = "reconnecting"
	SocketClosed       SocketState = "closed"
)

// SocketOptions configures a ChannelSocket.
type SocketOptions struct {
	// GetToken is called before every connection attempt (including
	// reconnects) so tokens are always fresh. Typically ChannelClient.
	// CreateToken, or a call to your own backend.
	GetToken func(ctx context.Context) (*TokenResponse, error)
	// URL is an explicit subscribe URL (wss://…/v1/channel/<instance>/subscribe)
	// when the token response's ws_url isn't used; ?token= is appended
	// automatically.
	URL string
	// Channels is the subset of the token's channels to subscribe on connect
	// (default: all).
	Channels []string
	// OnMessage receives every delivered message. Called from the socket's
	// read goroutine — hand off to your own goroutine for slow work.
	OnMessage func(ChannelMessage)
	// OnState observes lifecycle transitions.
	OnState func(SocketState)
	// OnError observes connection-level errors.
	OnError func(error)
	// PingInterval is the keepalive ping interval (default 30s). A silent
	// connection past ~1.5 intervals is treated as dead and reconnected.
	PingInterval time.Duration
	// BackoffBase/BackoffMax bound the reconnect backoff (defaults 250ms / 30s).
	BackoffBase time.Duration
	BackoffMax  time.Duration
}

type pendingSub struct {
	channels []string
	ch       chan subResult
}

type subResult struct {
	channels []string
	err      error
}

type pubResult struct {
	delivered int
	err       error
}

// ChannelSocket is a managed WebSocket subscriber for altengine channels:
// connect, auto-reconnect with backoff, token re-mint on reconnect,
// resubscribe, and keepalive pings.
type ChannelSocket struct {
	opts SocketOptions

	mu           sync.Mutex
	conn         *websocket.Conn
	state        SocketState
	desired      map[string]struct{}
	explicit     bool
	closed       bool
	attempts     int
	lastActivity time.Time
	pendingSubs  []pendingSub
	pendingPubs  []chan pubResult
	cancelLoop   context.CancelFunc
}

// NewChannelSocket builds a socket; call Connect to dial.
func NewChannelSocket(opts SocketOptions) *ChannelSocket {
	s := &ChannelSocket{opts: opts, state: SocketIdle, desired: map[string]struct{}{}}
	s.explicit = opts.Channels != nil
	for _, c := range opts.Channels {
		s.desired[c] = struct{}{}
	}
	if s.opts.PingInterval == 0 {
		s.opts.PingInterval = 30 * time.Second
	}
	if s.opts.BackoffBase == 0 {
		s.opts.BackoffBase = 250 * time.Millisecond
	}
	if s.opts.BackoffMax == 0 {
		s.opts.BackoffMax = 30 * time.Second
	}
	return s
}

// State returns the current lifecycle state.
func (s *ChannelSocket) State() SocketState {
	s.mu.Lock()
	defer s.mu.Unlock()
	return s.state
}

func (s *ChannelSocket) setState(st SocketState) {
	if s.state == st {
		return
	}
	s.state = st
	if s.opts.OnState != nil {
		go s.opts.OnState(st)
	}
}

func (s *ChannelSocket) emitError(err error) {
	if s.opts.OnError != nil {
		go s.opts.OnError(err)
	}
}

// Connect dials the socket; it returns once the connection is open. After a
// successful Connect the socket reconnects automatically until Close.
func (s *ChannelSocket) Connect(ctx context.Context) error {
	s.mu.Lock()
	if s.conn != nil {
		s.mu.Unlock()
		return nil
	}
	s.closed = false
	s.setState(SocketConnecting)
	s.mu.Unlock()

	if err := s.dial(ctx); err != nil {
		s.mu.Lock()
		s.setState(SocketClosed)
		s.mu.Unlock()
		return err
	}
	return nil
}

// Close stops reconnecting and closes the connection.
func (s *ChannelSocket) Close() error {
	s.mu.Lock()
	s.closed = true
	conn := s.conn
	s.conn = nil
	if s.cancelLoop != nil {
		s.cancelLoop()
	}
	s.failPendingLocked(errors.New("socket closed"))
	s.setState(SocketClosed)
	s.mu.Unlock()
	if conn != nil {
		return conn.Close(websocket.StatusNormalClosure, "client close")
	}
	return nil
}

// Subscribe adds channels (tracked and re-applied after reconnects) and waits
// for the server's ack, returning the acked channel list.
func (s *ChannelSocket) Subscribe(ctx context.Context, channels []string) ([]string, error) {
	s.mu.Lock()
	for _, c := range channels {
		s.desired[c] = struct{}{}
	}
	s.explicit = true
	conn := s.conn
	if conn == nil || s.state != SocketOpen {
		all := s.desiredLocked()
		s.mu.Unlock()
		return all, nil
	}
	ch := make(chan subResult, 1)
	s.pendingSubs = append(s.pendingSubs, pendingSub{channels: channels, ch: ch})
	s.mu.Unlock()

	if err := s.send(ctx, conn, map[string]any{"type": "subscribe", "channels": channels}); err != nil {
		return nil, err
	}
	select {
	case res := <-ch:
		return res.channels, res.err
	case <-ctx.Done():
		return nil, ctx.Err()
	}
}

// Unsubscribe removes channels from the subscription (and the resubscribe set).
func (s *ChannelSocket) Unsubscribe(ctx context.Context, channels []string) error {
	s.mu.Lock()
	for _, c := range channels {
		delete(s.desired, c)
	}
	s.explicit = true
	conn := s.conn
	open := s.state == SocketOpen
	s.mu.Unlock()
	if conn == nil || !open {
		return nil
	}
	return s.send(ctx, conn, map[string]any{"type": "unsubscribe", "channels": channels})
}

// Publish sends a message over the socket (requires a ws-capable publish
// token) and waits for the ack, returning the delivered count.
func (s *ChannelSocket) Publish(ctx context.Context, channel string, data any) (int, error) {
	s.mu.Lock()
	conn := s.conn
	if conn == nil || s.state != SocketOpen {
		s.mu.Unlock()
		return 0, errors.New("socket is not open")
	}
	ch := make(chan pubResult, 1)
	s.pendingPubs = append(s.pendingPubs, ch)
	s.mu.Unlock()

	if err := s.send(ctx, conn, map[string]any{"type": "publish", "channel": channel, "data": data}); err != nil {
		return 0, err
	}
	select {
	case res := <-ch:
		return res.delivered, res.err
	case <-ctx.Done():
		return 0, ctx.Err()
	}
}

// --- internals ---

func (s *ChannelSocket) desiredLocked() []string {
	out := make([]string, 0, len(s.desired))
	for c := range s.desired {
		out = append(out, c)
	}
	return out
}

func (s *ChannelSocket) send(ctx context.Context, conn *websocket.Conn, frame any) error {
	b, err := json.Marshal(frame)
	if err != nil {
		return err
	}
	return conn.Write(ctx, websocket.MessageText, b)
}

func (s *ChannelSocket) buildURL(tok *TokenResponse) (string, error) {
	raw := tok.WSURL
	if s.opts.URL != "" {
		raw = s.opts.URL
	}
	if raw == "" {
		return "", errors.New("no WebSocket URL: token response had no ws_url and no URL option was set")
	}
	u, err := url.Parse(raw)
	if err != nil {
		return "", err
	}
	q := u.Query()
	if q.Get("token") == "" {
		q.Set("token", tok.Token)
	}
	// Narrow the initial subscription to the explicitly requested subset.
	s.mu.Lock()
	if s.explicit && len(s.desired) > 0 {
		q.Set("channels", strings.Join(s.desiredLocked(), ","))
	}
	s.mu.Unlock()
	u.RawQuery = q.Encode()
	return u.String(), nil
}

func (s *ChannelSocket) dial(ctx context.Context) error {
	tok, err := s.opts.GetToken(ctx)
	if err != nil {
		return fmt.Errorf("get token: %w", err)
	}
	u, err := s.buildURL(tok)
	if err != nil {
		return err
	}
	conn, _, err := websocket.Dial(ctx, u, nil)
	if err != nil {
		return err
	}
	conn.SetReadLimit(1 << 20)

	loopCtx, cancel := context.WithCancel(context.Background())
	s.mu.Lock()
	s.conn = conn
	s.cancelLoop = cancel
	s.attempts = 0
	s.lastActivity = time.Now()
	s.setState(SocketOpen)
	s.mu.Unlock()

	go s.readLoop(loopCtx, conn)
	go s.pingLoop(loopCtx, conn)
	return nil
}

func (s *ChannelSocket) readLoop(ctx context.Context, conn *websocket.Conn) {
	for {
		_, data, err := conn.Read(ctx)
		if err != nil {
			s.onDisconnect(conn, err)
			return
		}
		s.mu.Lock()
		s.lastActivity = time.Now()
		s.mu.Unlock()
		s.handleFrame(data)
	}
}

func (s *ChannelSocket) handleFrame(data []byte) {
	raw := string(data)
	if raw == "pong" || raw == "ping" {
		return
	}
	var frame struct {
		Type      string          `json:"type"`
		Channels  []string        `json:"channels"`
		Channel   string          `json:"channel"`
		Data      json.RawMessage `json:"data"`
		TS        int64           `json:"ts"`
		Delivered int             `json:"delivered"`
		Error     json.RawMessage `json:"error"`
	}
	if err := json.Unmarshal(data, &frame); err != nil {
		return
	}
	switch frame.Type {
	case "subscribed":
		// Only settle the oldest pending subscribe if this ack covers its
		// channels — an unsolicited ack (e.g. for the connect-time ?channels=
		// subscription) arriving late must not steal a Subscribe's resolution.
		acked := make(map[string]struct{}, len(frame.Channels))
		for _, c := range frame.Channels {
			acked[c] = struct{}{}
		}
		s.mu.Lock()
		if len(s.pendingSubs) > 0 {
			head := s.pendingSubs[0]
			covered := true
			for _, c := range head.channels {
				if _, ok := acked[c]; !ok {
					covered = false
					break
				}
			}
			if covered {
				s.pendingSubs = s.pendingSubs[1:]
				head.ch <- subResult{channels: frame.Channels}
			}
		}
		s.mu.Unlock()
	case "published":
		s.mu.Lock()
		if len(s.pendingPubs) > 0 {
			ch := s.pendingPubs[0]
			s.pendingPubs = s.pendingPubs[1:]
			ch <- pubResult{delivered: frame.Delivered}
		}
		s.mu.Unlock()
	case "error":
		msg := strings.Trim(string(frame.Error), `"`)
		err := errors.New(msg)
		// An error ack settles the oldest pending request, if any.
		s.mu.Lock()
		if len(s.pendingPubs) > 0 {
			ch := s.pendingPubs[0]
			s.pendingPubs = s.pendingPubs[1:]
			ch <- pubResult{err: err}
		} else if len(s.pendingSubs) > 0 {
			head := s.pendingSubs[0]
			s.pendingSubs = s.pendingSubs[1:]
			head.ch <- subResult{err: err}
		}
		s.mu.Unlock()
		s.emitError(err)
	default:
		if frame.Type == "" && frame.Channel != "" {
			if s.opts.OnMessage != nil {
				s.opts.OnMessage(ChannelMessage{Channel: frame.Channel, Data: frame.Data, TS: frame.TS})
			}
		}
	}
}

func (s *ChannelSocket) pingLoop(ctx context.Context, conn *websocket.Conn) {
	ticker := time.NewTicker(s.opts.PingInterval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			s.mu.Lock()
			stale := time.Since(s.lastActivity) > s.opts.PingInterval*3/2
			s.mu.Unlock()
			if stale {
				// Dead connection: no traffic (not even pongs). Force a reconnect.
				conn.Close(websocket.StatusCode(4000), "keepalive timeout")
				return
			}
			if err := conn.Write(ctx, websocket.MessageText, []byte("ping")); err != nil {
				return
			}
		}
	}
}

func (s *ChannelSocket) onDisconnect(conn *websocket.Conn, cause error) {
	s.mu.Lock()
	if s.conn != conn {
		// A newer connection already replaced this one.
		s.mu.Unlock()
		return
	}
	s.conn = nil
	if s.cancelLoop != nil {
		s.cancelLoop()
	}
	s.failPendingLocked(fmt.Errorf("socket closed: %w", cause))
	if s.closed {
		s.mu.Unlock()
		return
	}
	s.setState(SocketReconnecting)
	attempt := s.attempts
	s.attempts++
	s.mu.Unlock()

	delay := s.opts.BackoffMax
	if attempt < 20 { // beyond 2^20 the shift only overflows; it's already capped
		delay = min(s.opts.BackoffBase<<attempt, s.opts.BackoffMax)
	}
	delay = time.Duration(float64(delay) * (0.5 + rand.Float64()*0.5))
	time.AfterFunc(delay, func() {
		s.mu.Lock()
		if s.closed || s.conn != nil {
			s.mu.Unlock()
			return
		}
		s.mu.Unlock()
		ctx, cancel := context.WithTimeout(context.Background(), 30*time.Second)
		defer cancel()
		if err := s.dial(ctx); err != nil {
			s.emitError(err)
			s.onDisconnect(nil, err)
		}
	})
}

func (s *ChannelSocket) failPendingLocked(err error) {
	for _, p := range s.pendingSubs {
		p.ch <- subResult{err: err}
	}
	s.pendingSubs = nil
	for _, ch := range s.pendingPubs {
		ch <- pubResult{err: err}
	}
	s.pendingPubs = nil
}

File: python/src/altengine/socket.py
"""ChannelSocket — the managed asyncio WebSocket subscriber for altengine
channels. Requires the ``ws`` extra (``pip install altengine[ws]``).

Your code supplies ``get_token`` (typically ``ChannelClient.create_token`` or a
call to your own backend); the socket does the rest — connect, auto-reconnect
with backoff, token re-mint on reconnect, resubscribe, and keepalive pings.
"""

from __future__ import annotations

import asyncio
import inspect
import json
import random
from typing import Any, Awaitable, Callable, Dict, List, Optional, Sequence, Union
from urllib.parse import parse_qs, urlencode, urlsplit, urlunsplit

TokenProvider = Callable[[], Union[Dict[str, Any], Awaitable[Dict[str, Any]]]]


class ChannelSocketError(Exception):
    """A server error frame or lifecycle failure on the socket."""


class ChannelSocket:
    """Managed WebSocket subscriber.

    ``get_token`` is called before every connection attempt (including
    reconnects) so tokens are always fresh; it returns the token-mint response
    (``{token, ws_url, ...}``). Messages arrive on :meth:`messages` (an async
    iterator) and/or the ``on_message`` callback.
    """

    def __init__(
        self,
        get_token: TokenProvider,
        url: Optional[str] = None,
        channels: Optional[Sequence[str]] = None,
        on_message: Optional[Callable[[Dict[str, Any]], None]] = None,
        ping_interval: float = 30.0,
        backoff_base: float = 0.25,
        backoff_max: float = 30.0,
    ) -> None:
        self._get_token = get_token
        self._url = url
        self._on_message = on_message
        self._ping_interval = ping_interval
        self._backoff_base = backoff_base
        self._backoff_max = backoff_max

        self.state = "idle"
        self._explicit = channels is not None
        self._desired: set = set(channels or [])
        self._ws: Any = None
        self._closed = False
        self._attempts = 0
        self._tasks: List[asyncio.Task] = []
        self._queue: asyncio.Queue = asyncio.Queue()
        self._pending_subs: List[tuple] = []  # (channels, future)
        self._pending_pubs: List[asyncio.Future] = []
        self._last_activity = 0.0

    async def connect(self) -> None:
        """Dial the socket; returns once the connection is open. After a
        successful connect the socket reconnects automatically until
        :meth:`close`."""
        if self._ws is not None:
            return
        self._closed = False
        self.state = "connecting"
        await self._dial()

    async def close(self) -> None:
        """Stop reconnecting and close the connection."""
        self._closed = True
        self.state = "closed"
        self._fail_pending(ChannelSocketError("socket closed"))
        for t in self._tasks:
            t.cancel()
        self._tasks = []
        ws, self._ws = self._ws, None
        if ws is not None:
            await ws.close()

    async def subscribe(self, channels: Sequence[str]) -> List[str]:
        """Subscribe to more channels (tracked and re-applied after
        reconnects); resolves with the server's ack."""
        self._desired.update(channels)
        self._explicit = True
        if self._ws is None or self.state != "open":
            return sorted(self._desired)
        fut: asyncio.Future = asyncio.get_running_loop().create_future()
        self._pending_subs.append((list(channels), fut))
        await self._ws.send(json.dumps({"type": "subscribe", "channels": list(channels)}))
        return await fut

    async def unsubscribe(self, channels: Sequence[str]) -> None:
        self._desired.difference_update(channels)
        self._explicit = True
        if self._ws is not None and self.state == "open":
            await self._ws.send(json.dumps({"type": "unsubscribe", "channels": list(channels)}))

    async def publish(self, channel: str, data: Any) -> int:
        """Publish over the socket (requires a ws-capable ``publish`` token);
        resolves with the delivered count from the server's ack."""
        if self._ws is None or self.state != "open":
            raise ChannelSocketError("socket is not open")
        fut: asyncio.Future = asyncio.get_running_loop().create_future()
        self._pending_pubs.append(fut)
        await self._ws.send(json.dumps({"type": "publish", "channel": channel, "data": data}))
        return await fut

    async def messages(self):
        """Async iterator over delivered messages (``{channel, data, ts}``)."""
        while not self._closed:
            msg = await self._queue.get()
            if msg is None:
                return
            yield msg

    # --- internals ---

    async def _token(self) -> Dict[str, Any]:
        tok = self._get_token()
        if inspect.isawaitable(tok):
            tok = await tok
        return tok

    def _build_url(self, tok: Dict[str, Any]) -> str:
        raw = self._url or tok.get("ws_url")
        if not raw:
            raise ChannelSocketError("no WebSocket URL: token response had no ws_url and no url was set")
        parts = urlsplit(raw)
        query = parse_qs(parts.query)
        if "token" not in query:
            query["token"] = [tok["token"]]
        # Narrow the initial subscription to the explicitly requested subset.
        if self._explicit and self._desired:
            query["channels"] = [",".join(sorted(self._desired))]
        return urlunsplit(parts._replace(query=urlencode(query, doseq=True)))

    async def _dial(self) -> None:
        import websockets

        tok = await self._token()
        ws = await websockets.connect(self._build_url(tok), max_size=1 << 20)
        self._ws = ws
        self._attempts = 0
        self._last_activity = asyncio.get_running_loop().time()
        self.state = "open"
        self._tasks = [
            asyncio.create_task(self._read_loop(ws)),
            asyncio.create_task(self._ping_loop(ws)),
        ]

    async def _read_loop(self, ws: Any) -> None:
        try:
            async for raw in ws:
                self._last_activity = asyncio.get_running_loop().time()
                self._handle_frame(raw)
        except Exception:
            pass
        finally:
            if self._ws is ws:
                await self._on_disconnect()

    def _handle_frame(self, raw: Any) -> None:
        if raw in ("ping", "pong"):
            return
        try:
            frame = json.loads(raw)
        except (ValueError, TypeError):
            return
        if not isinstance(frame, dict):
            return
        ftype = frame.get("type")
        if ftype == "subscribed":
            # Only settle the oldest pending subscribe if this ack covers its
            # channels — an unsolicited ack (e.g. for the connect-time
            # ?channels= subscription) must not steal a subscribe()'s resolution.
            acked = set(frame.get("channels") or [])
            if self._pending_subs and set(self._pending_subs[0][0]) <= acked:
                _, fut = self._pending_subs.pop(0)
                if not fut.done():
                    fut.set_result(frame.get("channels") or [])
        elif ftype == "published":
            if self._pending_pubs:
                fut = self._pending_pubs.pop(0)
                if not fut.done():
                    fut.set_result(frame.get("delivered", 0))
        elif ftype == "error":
            err = ChannelSocketError(str(frame.get("error")))
            # An error ack settles the oldest pending request, if any.
            if self._pending_pubs:
                fut = self._pending_pubs.pop(0)
                if not fut.done():
                    fut.set_exception(err)
            elif self._pending_subs:
                _, fut = self._pending_subs.pop(0)
                if not fut.done():
                    fut.set_exception(err)
        elif ftype is None and isinstance(frame.get("channel"), str):
            if self._on_message is not None:
                self._on_message(frame)
            self._queue.put_nowait(frame)

    async def _ping_loop(self, ws: Any) -> None:
        try:
            while True:
                await asyncio.sleep(self._ping_interval)
                now = asyncio.get_running_loop().time()
                if now - self._last_activity > self._ping_interval * 1.5:
                    # Dead connection: no traffic (not even pongs). Force a reconnect.
                    await ws.close(code=4000, reason="keepalive timeout")
                    return
                await ws.send("ping")
        except Exception:
            pass

    async def _on_disconnect(self) -> None:
        self._ws = None
        self._fail_pending(ChannelSocketError("socket closed"))
        if self._closed:
            return
        self.state = "reconnecting"
        delay = min(self._backoff_base * 2 ** min(self._attempts, 20), self._backoff_max)
        self._attempts += 1
        # Full jitter keeps concurrent reconnects from stampeding in unison.
        await asyncio.sleep(delay * (0.5 + random.random() * 0.5))
        if self._closed or self._ws is not None:
            return
        try:
            await self._dial()
        except Exception:
            await self._on_disconnect()

    def _fail_pending(self, err: Exception) -> None:
        for _, fut in self._pending_subs:
            if not fut.done():
                fut.set_exception(err)
        self._pending_subs = []
        for fut in self._pending_pubs:
            if not fut.done():
                fut.set_exception(err)
        self._pending_pubs = []
